### PR TITLE
Zoning E2E Test + E2E Improvements

### PR DESCRIPTION
**Pull request by ChristianAtDell**
# Description
The primary goal of this PR is to add an E2E test scenario (and robust testing script) that covers the new zoning work. In addition to that, some improvements were added to the E2E test suite.
1. All custom tests in E2E scenarios are now part of an 'array' of custom tests for their scenario, and can be selected individually. This allows for multiple custom tests to be run per scenario, which was necessary for adding and removing zone labels to k8s nodes. Future scenarios added will need to have their custom tests indented inwards to be part of an array, but legacy support for the previous 'run custom test' E2E step was left intact. 
2. A new way of deploying secrets as a part of E2E tests was introduced, that creates the secret by reading in the template file, editing the string in-memory, and deploying from that string. This was necessary because the previous way templates were used, editing in place, created an impossible scenario to restore from in our testing: if any two template keys had the same value, it was impossible to reverse the string replacement that was done. 

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| https://github.com/dell/csm/issues/1613 |

# Checklist:

- [X] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [X] I have verified that new and existing unit tests pass locally with my changes
- [X] I have not allowed coverage numbers to degenerate
- [X] I have maintained at least 80% code coverage
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have maintained backward compatibility
- [X] I have executed the relevant end-to-end test scenarios

# How Has This Been Tested?
E2E scenario has been run standalone. It passes, though that does not actually test zoning yet, as zoning work is not completed.

### CODE REVIEW

File: tests/e2e/modify_zoning_labels.sh [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+# To remove a zone label:
+# ./modify_zoning_labels.sh remove <zone-key>
+# To remove all zone labels:
+# ./modify_zoning_labels.sh remove-all-zones
+# To validate the zone labels:
+# ./modify_zoning_labels.sh validate-zoning
+
+# get all worker node names in the cluster
+get_worker_nodes() {
+  kubectl get nodes -A | grep -v -E 'mast.r|control-plane'  | grep -v NAME | awk '{ print $1 }'
```
Comment (donatwork) [inline tests/e2e/modify_zoning_labels.sh:29]: Is mast.r intentional?

Comment (shaynafinocchiaro) [reply tests/e2e/modify_zoning_labels.sh]: Fixed

File: tests/e2e/steps/steps_def.go [hunk 2]
```diff
@@ -752,6 +756,46 @@ func (step *Step) setUpSecret(res Resource, templateFile, name, namespace, crTyp
...
+	if secretExists(namespace, name) {
+		err := execCommand("kubectl", "delete", "secret", "-n", namespace, name)
+		if err != nil {
+			return fmt.Errorf("failed to delete secret: %s", err.Error())
+		}
+	}
+
+	// create new secret
+	fileArg := "--from-literal=config=" + fileString
+	err = execCommand("kubectl", "create", "secret", "generic", "-n", namespace, name, fileArg)
```
Comment (donatwork) [inline tests/e2e/steps/steps_def.go:791]: Have you tested this with special shell characters in fileString, e.g. dollar signs and exclamation marks.

**Review by anathoodell — APPROVED** [review]

File: tests/e2e/modify_zoning_labels.sh [hunk 3]
```diff
@@ -0,0 +1,197 @@
+# Copyright © 2024 Dell Inc. or its subsidiaries. All Rights Reserved.
```
Comment (JacobGros) [inline tests/e2e/modify_zoning_labels.sh:1]: Should we validate SDC connectivity to array(s) as well as part of this script?
That could also come later as part of the next story

**Review by JacobGros — APPROVED** [review]

**Review by donatwork — APPROVED** [review]

File: tests/e2e/modify_zoning_labels.sh (hunk 3 above)
Comment (shaynafinocchiaro) [reply tests/e2e/modify_zoning_labels.sh:1]: @JacobGros can we get this information through driver pods or nodes on the cluster?

Comment (ChristianAtDell) [reply tests/e2e/modify_zoning_labels.sh:1]: We've looked into this, but the system running the e2e tests will almost assuredly *not* be the worker node and the only way to perform this test would be to SSH into the worker node as a part of the test-- I do not recommend this.

**Review by atye — APPROVED** [review]

**Comment by ChristianAtDell** [comment]
Ignore the commit to revert, I went to click 'delete branch' and hit 'revert' instead, which opened a new branch with a revert commit. I've deleted the branch.

**Pull request merged**